frontend/src/routes/password.change.tsx

```diff
@@ -0,0 +1,283 @@
+            </Form.Label>
+
+            <Form.PasswordControl
+              required
+              autoComplete="new-password"
+              ref={newPasswordRef}
+              onBlur={() =>
+                newPasswordAgainRef.current!.value &&
+                newPasswordAgainRef.current!.reportValidity()
+              }
+            />
+
+            {/* TODO Show a password bar. https://github.com/matrix-org/matrix-authentication-service/issues/2854 */}
+
+            <Form.ErrorMessage match="valueMissing">
+              {t("frontend.errors.field_required")}
+            </Form.ErrorMessage>
+
+            {result.data &&
+              result.data.setPassword.status ===
+                SetPasswordStatus.InvalidNewPassword && (
+                <Form.ErrorMessage>
+                  {t(
+                    "frontend.password_change.failure.description.invalid_new_password",
+                  )}
+                </Form.ErrorMessage>
+              )}
+          </Form.Field>
+
+          <Form.Field name="new_password_again">
+            {/*
+            TODO This field has validation defects,
+            some caused by Radix-UI upstream bugs.
+            https://github.com/matrix-org/matrix-authentication-service/issues/2855
+          */}
+            <Form.Label>
+              {t("frontend.password_change.new_password_again_label")}
+            </Form.Label>
+
+            <Form.PasswordControl
+              required
+              ref={newPasswordAgainRef}
+              autoComplete="new-password"
+            />
+
+            <Form.ErrorMessage match="valueMissing">
+              {t("frontend.errors.field_required")}
+            </Form.ErrorMessage>
+
+            <Form.ErrorMessage
+              match={(v, form) => v !== form.get("new_password")}
+            >
+              {t("frontend.password_change.passwords_no_match")}
+            </Form.ErrorMessage>
+
+            <Form.HelpMessage match="valid">
+              {/* TODO Use SuccessMessage once ready. https://github.com/matrix-org/matrix-authentication-service/issues/2856 */}
+              <IconCheckCircleSolid />
+              {t("frontend.password_change.passwords_match")}
+            </Form.HelpMessage>
+          </Form.Field>
+
+          {unhandleableError && (
+            <Alert
+              type="critical"
+              title={t("frontend.password_change.failure.title")}
+            >
+              {t("frontend.password_change.failure.description.unspecified")}
+            </Alert>
+          )}
+
+          <Form.Submit kind="primary" disabled={result.fetching}>
+            {!!result.fetching && <LoadingSpinner inline />}
+            {t("action.save")}
+          </Form.Submit>
+
+          {handleableError && (
+            <Alert
+              type="critical"
+              title={t("frontend.password_change.failure.title")}
+            >
+              {errorMsg}
+            </Alert>
+          )}
```

Comment:
I'm not sure I like the `switch/case` logic up there. Because we have some errors specific to fields, I would rather have a series of `{result.data?.setPassword.status === SetPasswordStatus.NoCurrentPassword && (<Alert … />)}`. wdyt?

Also could you stick alerts at the top of the form?

Reply:
do you think we should not show the errors in an alert if they are specific to a form field? I was thinking that might make it a bit harder to notice but wasn't entirely sure.

Both the switch and the sequence of conditionally rendered alerts feel a bit tedious but struggling to think of anything better

Reply:
that said: I think it's important to have a 'default' case (for unhandled errors etc) and by taking your approach, it's harder to see how we'd do that

Reply:
> do you think we should not show the errors in an alert if they are specific to a form field? I was thinking that might make it a bit harder to notice but wasn't entirely sure.

Yep, as radix forms will take care of focusing the errored field. I don't think I've saw any form with an alert like that in the designs

> that said: I think it's important to have a 'default' case (for unhandled errors etc) and by taking your approach, it's harder to see how we'd do that

that's a fair point. Although could be offset by doing something like:

```typescript
  const status = result.data?.setPassword.status;
  switch (status) {
    // Cases that are handled by the form
    case SetPasswordStatus.InvalidNewPassword:
    case SetPasswordStatus.NoCurrentPassword:
    case SetPasswordStatus.PasswordChangesDisabled:
    case SetPasswordStatus.WrongPassword:
    case SetPasswordStatus.Allowed:
    case undefined:
      break;

    case SetPasswordStatus.NotFound:
    case SetPasswordStatus.NotAllowed:
      throw new Error(`unexpected error when changing password: ${status}`);

    default:
      unreachable(status);
  }
```